test/models/host_test.rb

```diff
@@ -44,6 +44,24 @@ class HostTest < ActiveSupport::TestCase
     assert host.save
   end
 
+  test "host without primary interface should rise error" do
+    host = FactoryBot.build(:host, :managed)
+    host.interfaces = [] # remove existing primary interface
+    host.interfaces = [ FactoryBot.create(:nic_managed, :primary => false, :provision => true, :host => host,
+                                          :domain => FactoryBot.create(:domain)) ]
+    refute host.valid?, "An interface marked as primary is missing"
+    assert_equal [:interfaces, :base], host.errors.keys
+  end
+
+  test "host without primary and provision interface should rise error" do
+    host = FactoryBot.build(:host, :managed)
+    host.interfaces = [] # remove existing primary interface
+    host.interfaces = [ FactoryBot.create(:nic_managed, :primary => false, :provision => false, :host => host,
+                                          :domain => FactoryBot.create(:domain)) ]
+    refute host.valid?, "An interface marked as primary is missing"
+    assert_equal [:interfaces, :base], host.errors.keys
```

Comment:
Can you also assert messages? Just for future regressions. Here and on line 52.

Reply:
Added error message comparison